### ISSUE TITLE
Request to get DLT_OPENFLOW ID for OpenBSD

### ISSUE THREAD

**Issue by reyk** [the-tcpdump-group/libpcap#542]
Hi,

TL;DR we need an official ID for DLT_OPENFLOW that is used in OpenBSD. For the time being, we are using DLT_USER0 for it.

Background: OpenBSD has a new in-kernel switch implementation that uses OpenFlow to communicate with the userland: http://man.openbsd.org/OpenBSD-current/man4/switch.4

I added a new data link type DLT_OPENFLOW to monitor the OpenFlow messages that are exchanged with the switch interface and it is supported by OpenBSD's version of libpcap and tcpdump.

The messages of this link type are raw OpenFlow messages prepended with a 32 bit field (in network byte order) to indicate the direction: controller to switch (1), switch to controller (2). UPDATE: prepended by the 64 bit data path id. And, yes, a version number can be encoded in the 32 bit direction field in the future, if ever needed.

```
/* Used by the bpf for DLT_OPENFLOW */
struct dlt_openflow_hdr {
	uint32_t	of_direction;
	uint64_t	of_datapath_id;
} __packed;

#define DLT_OPENFLOW_TO_SWITCH		1
#define DLT_OPENFLOW_TO_CONTROLLER	2
```

Reyk

**Comment by infrastation** [comment]
Do you know why this kernel-mode switch does not communicate with the controller using a normal TCP connection?

**Comment by infrastation** [comment]
It's not that important perhaps -- you can always fit a coupling userland process between the character device and a TCP endpoint.

**Comment by infrastation** [comment]
(In which case it will be simpler to capture the translated TCP session, but those interpretations don't seem to be quite relevant in the scope of this request.)

**Comment by reyk** [comment]
> Do you know why this kernel-mode switch does not communicate with the controller using a normal TCP connection?

Because we designed it that way - we're not doing TCP between kernel and userland.  The control channel is a simple character device that exchanges raw openflow messages with the small header, there is no need to use TCP here. In fact, the message-oriented OpenFlow is not restricted to TCP or TLS. It also works over UDP (only a few implementation support that). And, yes, a userland program like our switchd can open the device and proxy the messages to one or more remote controllers.

- The userland opens /dev/switchX to communicate with the related switchX interface
- The bpf listener is a way for out-of-band monitoring of the communication.
- We have similar concepts for DLT_PFLOG, DLT_PFSYNC, and DLT_ENC in OpenBSD already.

```
# tcpdump -L -i switch0            
2 link types supported:
        OPENFLOW
        EN10MB
# tcpdump -y openflow -i switch0 
tcpdump: listening on switch0, link-type OPENFLOW
```

OpenBSD's version of libpcap and tcpdump have diverged quite a bit from upstream, but that shouldn't hinder the tcpdump group from assigning us that ID ;-)

So could we have a somewhat official number for DLT_OPENFLOW, please?

**Comment by guyharris** [comment]
So is there a *stable* specification for the format of `DLT_OPENFLOW` packets, i.e. something better than "include this OpenBSD header file on your platform" (which is the "specification" for `DLT_PFLOG`)?  "Stable" meaning "something that's not going to change from OpenBSD release to OpenBSD release or, if you can't guarantee that, something that at least has a version number that will change if the format changes".

**Comment by reyk** [comment]
Why does this even matter? Neither OpenBSD's DLT_PFLOG nor DLT_ENC have a version field and they're also only used between kernel and user land. We have a version field in PFSYNC because the protocol is sent over the wire and the keeping it in sync is somewhat important.

Here is the header that will be included in OpenBSD 6.1 (struct dlt_openflow_hdr): https://github.com/openbsd/src/blob/07d35acad2a058f859b960a3964765d7a9bca875/sys/net/ofp.h#L889  - Any future versions can use the reserved bits in the 32bit direction field, so encoding a version number is certainly possible.

And, btw., I count 23 entries for DLT_JUNIPER_*, so what's the point of not giving us just another single ID right away?

**Comment by guyharris** [comment]
> Why does this even matter?

So that we know whether to say, in the entry for it on [the link-layer header types page](http://www.tcpdump.org/linktypes.html), "here's a specification" or something more like what we say for `DLT_PFLOG`:

> OpenBSD pflog; the link-layer header contains a "struct pfloghdr" structure, as defined by the host on which the file was saved. (This differs from operating system to operating system and release to release; there is nothing in the file to indicate what the layout of that structure is.)

**Comment by reyk** [comment]
Last change of DLT_PFLOG was over 6 years ago.

**Comment by guyharris** [comment]
> Last change of DLT_PFLOG was over 6 years ago.

Yes, but it still changed, and does not rule out future changes.

If you want a DLT_/LINKTYPE_ value for an OS-internal message format that's *subject* to change, regardless of how often it changes, and has no versioning mechanism, I'm willing to assign one, but I'm *not* going to give any more of a description that one like the one for `DLT_PFLOG` or to devote any energy to dealing with dissector incompatibilities between versions (e.g., developing heuristics to try to detect the format).

I've assigned 267 to DLT_OPENFLOW.

**Comment by reyk** [comment]
Thanks

**Comment by guyharris** [comment]
You're welcome.

In return, could you please request a `DLT_IEEE802_11_RADIO_OPENBSD` link-layer header type, and use it in OpenBSD, given that OpenBSD assign different meanings to the fields with presence bits 14 (FCS vs. RX_FLAGS), 15 (HWQUEUE vs. TX_FLAGS), and 16 (RSSI vs. RTS_RETRIES), so that software that interprets `DLT_IEEE802_11_RADIO` correctly in files from other platforms would interpret Radiotap files from OpenBSD incorrectly?  If OpenBSD were to use a different DLT_ value, software could interpret that DLT_ value correctly for OpenBSD.

**Comment by reyk** [comment]
I forwarded this to stsp@, who is taking care of net80211 these days.

Radiotap used to be a standard that was portable. I added the fields in OpenBSD _before_ they were added to the Linux kernel and now you're asking us to change the DLT? That doesn't make sense. I think Linux should have looked at the other implementations first before assuming world dominance. Remember, Radiotap came from BSD.

https://github.com/openbsd/src/blame/master/sys/net80211/ieee80211_radiotap.h#L184
https://github.com/torvalds/linux/blame/master/include/net/ieee80211_radiotap.h#L213

**Comment by guyharris** [comment]
A history of various flavors of `ieee80211_radiotap.h` from various CVS/SVN/Git repositories:

This change in FreeBSD:

    r140630 | sam | 2005-01-22 12:12:05 -0800 (Sat, 22 Jan 2005) | 10 lines

    o replace out-of-line copy of FCS w/ a flag that indicates the
      frame includes FCS (requires applications to be updated, but since
      we weren't doing the out-of-line FCS stuff anyway app changes
      were needed already)
    o add a flag to indicate padding exists between the 802.11 header and
      the payload (e.g. for Atheros cards)
    o diff reducation against netbsd

removed the definition of `IEEE80211_RADIOTAP_FCS` as 14.

This change in NetBSD:

    revision 1.11
    date: 2005-06-22 06:16:02 +0000;  author: dyoung;  state: Exp;  lines: +9 -5;
    branches:  1.11.2;
    Resolve conflicts in importation of 18-May-2005 ath(4) / net80211(9)
    from FreeBSD.  Introduce compatibility shims (sys/dev/ic/ath_netbsd.[ch],
    sys/net80211/ieee80211_netbsd.[ch]).  Update drivers (an, atu, atw,
    awi, ipw, iwi, rtw, wi) for the new net80211(9) API.

propagated that removal to NetBSD.

This change in FreeBSD:

    r170530 | sam | 2007-06-10 20:36:55 -0700 (Sun, 10 Jun 2007) | 55 lines

    Update 802.11 wireless support:
        [long description elided]

added `IEEE80211_RADIOTAP_XCHANNEL` as 14.

This change in FreeBSD:

    r171127 | thompsa | 2007-06-30 23:59:30 -0700 (Sat, 30 Jun 2007) | 6 lines

    Move the XCHANNEL radiotap definition to a free slot.  NetBSD defined some
    additional values causing a conflict.

    Submitted by:   sam
    Approved by:    re (kensmith)

moved `IEEE80211_RADIOTAP_XCHANNEL` to 18 to leave a "gap for netbsd definitions".

This change in NetBSD:

    revision 1.17
    date: 2007-03-26 04:32:14 +0000;  author: dyoung;  state: Exp;  lines: +37 -1;
    Define four new radiotap fields per discussions with John Bicket, Sam
    Leffler, and others:
        
            IEEE80211_RADIOTAP_RX_FLAGS = 14,
            IEEE80211_RADIOTAP_TX_FLAGS = 15,
            IEEE80211_RADIOTAP_RTS_RETRIES = 16,
            IEEE80211_RADIOTAP_DATA_RETRIES = 17,
        
    I describe the fields in the manual page and in comments in the
    header file (cross-referenced by the manual page).

Introduced the fields in question.

This change in FreeBSD:

    r245156 | adrian | 2013-01-07 22:59:21 -0800 (Mon, 07 Jan 2013) | 3 lines

    Add in the missing radiotap definitions from the sipsolutions.net
    radiotap "upstream" source.

added the four fields in question.

This change in Linux:

    commit 876c9d3aeb989cf1961f2c228d309ba5dcfb1172
    Author: Marcelo Tosatti <marcelo@kvack.org>
    Date:   Sat Feb 10 12:25:27 2007 -0200

        [PATCH] Marvell Libertas 8388 802.11b/g USB driver

        Add the Marvell Libertas 8388 802.11 USB driver.

        Signed-off-by: Marcelo Tosatti <marcelo@kvack.org>
        Signed-off-by: John W. Linville <linville@tuxdriver.com>

introduced those fields to Linux, so, unless the Linux Git repository has an incorrect date for the change in question, it wasn't Linux that first had those fields, it was NetBSD.

Had OpenBSD picked up NetBSD's revisions 1.10 and 1.17, and collaborated with other radiotap developers when adding additional fields, there would be no compatibility issues here; it's OpenBSD that's the odd man out.

So, yes, I *am* asking you to change the DLT_, so you don't differ incompatibly from NetBSD, FreeBSD, and Linux (and possibly macOS, although Apple don't publish that header)

**Comment by reyk** [comment]
> Had OpenBSD picked up NetBSD's revisions 1.10 and 1.17, and collaborated with other radiotap developers when adding additional fields, there would be no compatibility issues here; it's OpenBSD that's the odd man out.

We missed one bit, the removal of the FCS, but our changes have been done in 2005/07/30.

You seem to lack the context: FreeBSD and NetBSD didn't want to collaborate with us at the time. They were deeply offended that I implemented the ar5k "OpenHAL" - which got ported to Linux later - because they had the binary drivers that were contract work by sam@freebsd instead. At least for FreeBSD, times have changed, sam retired, and we're having a better collaboration in wireless now.

Anyway, enough OS bashing here, we might just sync the IDs from the other BSDs but I doubt that they will integrate our extensions, even after moving them to unused IDs, but we can give it a try. It doesn't make much sense to use a different Radiotap DLT here.

**Comment by infrastation** [comment]
@reyk, if you would be interested in some more input as to why it sometimes helps to have a protocol structure defined in a standalone spec (possibly with a version field), it is because the packet capture may be written to a file on one host at one time and decoded on another host and/or at a very different time. Under-defined protocols have happened in the past and caused unnecessary difficulties, which itself cannot be fixed, but Guy tried to steer away from it in a new design as much as seemed possible.

You may also want to look how tcpdump decodes OpenFlow version 1.0, which itself is a good example of protocol versioning.

Thanks.

**Comment by reyk** [comment]
Makes sense. I will update the dlt_openflow_hdr header above to include a version field, it is already reserved in the 32 bit direction field. So it should just work(TM) if we consider the current version as version 0.

We cannot really fix the pflog situation at this point. The problem is that FreeBSD and derivates like macOS still use a PF version from more than 6 or 7 years ago, so they never upgraded to the current PF and pflog from OpenBSD. I also don't think that we can add a version number here. As for Radiotap, we will just try to sync it.

The PCAP format itself is kind of set in stone, but everything would have been easier with a OS id in the file header (much like in ELF binaries), as many differences seem to be related to the OS where it was captured.
